linter/lints/crl/lints.go

```diff
@@ -369,3 +372,135 @@ func hasMozReasonCodes(crl *crl_x509.RevocationList) *lint.LintResult {
+//
+// GeneralizedTime values MUST be expressed in Greenwich Mean Time (Zulu) and
+// MUST include seconds (i.e., times are YYYYMMDDHHMMSSZ), even where the number
+// of seconds is zero.  GeneralizedTime values MUST NOT include fractional
+// seconds. See: https://www.rfc-editor.org/rfc/rfc5280.html#section-4.1.2.5.2
+func hasValidTimestamps(crl *crl_x509.RevocationList) *lint.LintResult {
+	input := cryptobyte.String(crl.RawTBSRevocationList)
+	parseFail := lint.LintResult{
+		Status:  lint.Error,
+		Details: "Failed to re-parse tbsCertList during linting",
+	}
+	// Read tbsCertList.
+	var tbs cryptobyte.String
+	input.ReadASN1(&tbs, cryptobyte_asn1.SEQUENCE)
+	// Skip (optional) version.
+	if !tbs.SkipOptionalASN1(cryptobyte_asn1.INTEGER) {
+		return &parseFail
+	}
+	// Skip signature.
+	if !tbs.SkipASN1(cryptobyte_asn1.SEQUENCE) {
+		return &parseFail
+	}
+	// Skip issuer.
+	if !tbs.SkipASN1(cryptobyte_asn1.SEQUENCE) {
+		return &parseFail
+	}
+	// Read thisUpdate.
+	var thisUpdate cryptobyte.String
+	var thisUpdateTag cryptobyte_asn1.Tag
+	if !tbs.ReadAnyASN1Element(&thisUpdate, &thisUpdateTag) {
+		return &parseFail
+	}
+	// Lint thisUpdate.
+	err := lintTimestamp(&thisUpdate, thisUpdateTag)
+	if err != nil {
+		parseFail.Details = err.Error()
+		return &parseFail
+	}
+	// Peek (optional) nextUpdate.
+	if tbs.PeekASN1Tag(cryptobyte_asn1.UTCTime) || tbs.PeekASN1Tag(cryptobyte_asn1.GeneralizedTime) {
+		// Read nextUpdate.
+		var nextUpdate cryptobyte.String
+		var nextUpdateTag cryptobyte_asn1.Tag
+		if !tbs.ReadAnyASN1Element(&nextUpdate, &nextUpdateTag) {
+			return &parseFail
+		}
+		// Lint nextUpdate.
+		err = lintTimestamp(&nextUpdate, nextUpdateTag)
+		if err != nil {
+			parseFail.Details = err.Error()
+			return &parseFail
+		}
+	}
+	// Read sequence of revokedCertificate.
+	var revokedSeq cryptobyte.String
+	tbs.ReadASN1(&revokedSeq, cryptobyte_asn1.SEQUENCE)
+	for !revokedSeq.Empty() {
+		// Read revokedCertificate.
+		var certSeq cryptobyte.String
+		revokedSeq.ReadASN1Element(&certSeq, cryptobyte_asn1.SEQUENCE)
+		certSeq.ReadASN1(&certSeq, cryptobyte_asn1.SEQUENCE)
+
+		// Skip userCertificate (serial number).
+		certSeq.SkipASN1(cryptobyte_asn1.INTEGER)
+
+		// Read revocationDate.
+		var revocationDate cryptobyte.String
+		var revocationDateTag cryptobyte_asn1.Tag
+		if !certSeq.ReadAnyASN1Element(&revocationDate, &revocationDateTag) {
+			return &parseFail
+		}
+
+		// Lint revocationDate.
+		err = lintTimestamp(&revocationDate, revocationDateTag)
+		if err != nil {
+			parseFail.Details = err.Error()
+			return &parseFail
+		}
+	}
+	return &lint.LintResult{Status: lint.Pass}
+}
+
+func lintTimestamp(der *cryptobyte.String, tag cryptobyte_asn1.Tag) error {
+	utcTimeFormat := "YYMMDDHHMMSSZ"
+	generalizedTimeFormat := "YYYYMMDDHHMMSSZ"
+	// Preserve the original timestamp for length checking.
+	derBytes := *der
+	var tsBytes cryptobyte.String
+	derBytes.ReadASN1(&tsBytes, tag)
+	tsString := string(tsBytes)
```

Comment:
I'd suggest `tsLen := len(string(tsBytes))` so that it's clearer that we only care about this for the sake of its length; we aren't doing any other complex processing on it.